File: Wire-iOS/Sources/UserInterface/Registration/NoHistoryViewController.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "NoHistoryViewController.h"
@import PureLayout;
#import "WAZUIMagicIOS.h"
#import "UIFont+MagicAccess.h"
#import "RegistrationFormController.h"
#import "Button.h"


@interface NoHistoryViewController ()
@property (nonatomic) UILabel *heroLabel;
@property (nonatomic) UILabel *subtitleLabel;
@property (nonatomic) UIView *contentView;
@property (nonatomic) Button *OKButton;
@end

@implementation NoHistoryViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.view.backgroundColor = [UIColor clearColor];
    
    [self createContentView];
    [self createHeroLabel];
    [self createSubtitleLabel];
    [self createOKButton];
    
    [self createViewConstraints];
    
    // Layout first to avoid the initial layout animation during the presentation. 
    [self.view layoutIfNeeded];
}

- (void)createContentView
{
    self.contentView = [[UIView alloc] initForAutoLayout];
    self.contentView.backgroundColor = [UIColor clearColor];
    [self.view addSubview:self.contentView];
}

- (void)createHeroLabel
{
    self.heroLabel = [[UILabel alloc] initForAutoLayout];
    self.heroLabel.numberOfLines = 0;
    
    NSMutableParagraphStyle *paragraphStyle = [[NSParagraphStyle defaultParagraphStyle] mutableCopy];
    paragraphStyle.paragraphSpacing = 10;
    NSString *heroText = [self localizableStringForPart:@"hero"];
    NSMutableAttributedString *attributedText = [[NSMutableAttributedString alloc] initWithString:NSLocalizedString(heroText, nil)
                                                                                       attributes:@{ NSParagraphStyleAttributeName : paragraphStyle,
                                                                                                     NSFontAttributeName: [UIFont fontWithMagicIdentifier:@"style.text.large.font_spec_medium"]}];
    self.heroLabel.attributedText = attributedText;
    
    [self.contentView addSubview:self.heroLabel];
}

- (void)createSubtitleLabel
{
    self.subtitleLabel = [[UILabel alloc] initForAutoLayout];
    self.subtitleLabel.numberOfLines = 0;
    NSMutableParagraphStyle *paragraphStyle = [[NSParagraphStyle defaultParagraphStyle] mutableCopy];
    paragraphStyle.paragraphSpacing = 10;
    NSString *subtitleText = [self localizableStringForPart:@"subtitle"];
    NSMutableAttributedString *attributedText = [[NSMutableAttributedString alloc] initWithString:NSLocalizedString(subtitleText, nil)
                                                                                       attributes:@{ NSParagraphStyleAttributeName : paragraphStyle,
                                                                                                     NSFontAttributeName: [UIFont fontWithMagicIdentifier:@"style.text.large.font_spec_thin"]}];
    self.subtitleLabel.attributedText = attributedText;
    [self.contentView addSubview:self.subtitleLabel];
}

- (void)createOKButton
{
    self.OKButton = [Button buttonWithStyle:ButtonStyleFullMonochrome];
    self.OKButton.translatesAutoresizingMaskIntoConstraints = NO;
    NSString *gotItText = [self localizableStringForPart:@"got_it"];
    [self.OKButton setTitle:NSLocalizedString(gotItText, nil) forState:UIControlStateNormal];
    [self.OKButton addTarget:self action:@selector(okButtonPressed:) forControlEvents:UIControlEventTouchUpInside];
    
    [self.contentView addSubview:self.OKButton];
}

- (void)okButtonPressed:(id)sender
{
    [self.formStepDelegate didCompleteFormStep:self];
}

- (void)createViewConstraints
{
    [self.heroLabel autoPinEdgeToSuperviewEdge:ALEdgeLeft withInset:28];
    [self.heroLabel autoPinEdgeToSuperviewEdge:ALEdgeRight withInset:28];
    
    [self.subtitleLabel autoPinEdgeToSuperviewEdge:ALEdgeLeft withInset:28];
    [self.subtitleLabel autoPinEdgeToSuperviewEdge:ALEdgeRight withInset:28];
    [self.subtitleLabel autoPinEdge:ALEdgeTop toEdge:ALEdgeBottom ofView:self.heroLabel withOffset:12];
    
    [self.OKButton autoPinEdge:ALEdgeTop toEdge:ALEdgeBottom ofView:self.subtitleLabel withOffset:24];
    [self.OKButton autoPinEdgesToSuperviewEdgesWithInsets:UIEdgeInsetsMake(0, 28, 28, 28) excludingEdge:ALEdgeTop];
    [self.OKButton autoSetDimension:ALDimensionHeight toSize:40];
    
    if (self.traitCollection.userInterfaceIdiom == UIUserInterfaceIdiomPad) {
        [self.contentView autoSetDimension:ALDimensionWidth toSize:self.parentViewController.maximumFormSize.width];
        [self.contentView autoSetDimension:ALDimensionHeight toSize:self.parentViewController.maximumFormSize.height];
        [self.contentView autoCenterInSuperview];
    } else {
        [self.contentView autoPinEdgesToSuperviewEdgesWithInsets:UIEdgeInsetsZero];
    }
}


- (NSString *)localizableStringForPart:(NSString *)part
{
    switch (self.contextType) {
        case ContextTypeNewDevice:
            return [@"registration.no_history" stringByAppendingPathExtension:part];
            break;
        case ContextTypeLoggedOut:
            return [@"registration.no_history.logged_out" stringByAppendingPathExtension:part];
            break;
    }
}

@end

File: Wire-iOS/Sources/UserInterface/Registration/RegistrationPersonal/Steps/NoHistoryViewController.swift
//
// Wire
// Copyright (C) 2018 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import Foundation
import WireDataModel


extension NoHistoryViewController {
    static let WireBackupUTI = "com.wire.backup-ios"
    
    @objc public func createButtons() {
        let restoreBackupButton = Button(style: .emptyMonochrome)
        restoreBackupButton.translatesAutoresizingMaskIntoConstraints = false
        restoreBackupButton.setTitle("registration.no_history.restore_backup".localized.uppercased(),
                                     for: .normal)
        
        restoreBackupButton.addCallback(for: .touchUpInside) { [unowned self] _ in
            if self.contextType == .loggedOut {
                self.showWarningMessage()
            }
            else {
                self.showFilePicker()
            }
        }
        
        restoreBackupButton.heightAnchor.constraint(equalToConstant: 40).isActive = true
        stackView.addArrangedSubview(restoreBackupButton)
    
    
        let okButton = Button(style: .fullMonochrome)
        okButton.translatesAutoresizingMaskIntoConstraints = false
        let gotItText = self.localizableString(forPart: "got_it")!
        okButton.setTitle(gotItText.localized, for: .normal)
        okButton.addCallback(for: .touchUpInside) { [unowned self] _ in
            self.showLoadingView = true
            self.formStepDelegate.didCompleteFormStep(self)
        }
        
        okButton.heightAnchor.constraint(equalToConstant: 40).isActive = true
        stackView.addArrangedSubview(okButton)
    }
    
    @objc public func createContentViewConstraints() {
        if self.traitCollection.horizontalSizeClass == .regular {
            NSLayoutConstraint.activate([
                contentView.widthAnchor.constraint(equalToConstant: parent!.maximumFormSize.width),
                contentView.heightAnchor.constraint(equalToConstant: parent!.maximumFormSize.height),
                contentView.centerXAnchor.constraint(equalTo: view.centerXAnchor),
                contentView.centerYAnchor.constraint(equalTo: view.centerYAnchor)
            ])
        }
        else {
            contentView.fitInSuperview()
        }
        
        NSLayoutConstraint.activate([
            stackView.leadingAnchor.constraint(equalTo: contentView.leadingAnchor, constant: 28),
            stackView.bottomAnchor.constraint(equalTo: contentView.bottomAnchor, constant: -28 - UIScreen.safeArea.bottom),
            stackView.trailingAnchor.constraint(equalTo: contentView.trailingAnchor, constant: -28)
        ])
    }

    fileprivate func showWarningMessage() {
        let alert = UIAlertController(title: "registration.no_history.restore_backup_warning.title".localized,
                                      message: "registration.no_history.restore_backup_warning.message".localized,
                                      cancelButtonTitle: "general.cancel".localized)
        
        let proceedAction = UIAlertAction(title: "registration.no_history.restore_backup_warning.proceed".localized,
                                          style: .default) { [unowned self] _ in
             self.showFilePicker()
        }
        alert.addAction(proceedAction)
        
        self.present(alert, animated: true)
    }
    
    fileprivate func showFilePicker() {
        // Test code to verify restore
        #if arch(i386) || arch(x86_64)
            let testFilePath = "/var/tmp/backup.ios_wbu"
            if FileManager.default.fileExists(atPath: testFilePath) {
                self.restore(with: URL(fileURLWithPath: testFilePath))
                return
            }
        #endif
        
        let picker = UIDocumentPickerViewController(documentTypes: [NoHistoryViewController.WireBackupUTI],
                                                  in: .`import`)
        picker.delegate = self
        self.present(picker, animated: true)
    }
    
    private func errorMessage(for error: Error) -> String {
        switch error {
        case StorageStack.BackupImportError.incompatibleBackup(let underlyingError):
            switch underlyingError {
            case BackupMetadata.VerificationError.backupFromNewerAppVersion:
                return "registration.no_history.restore_backup_failed.wrong_version.message".localized
            case BackupMetadata.VerificationError.userMismatch:
                return "registration.no_history.restore_backup_failed.wrong_account.message".localized
            default:
                return "registration.no_history.restore_backup_failed.message".localized
            }
            
        default:
            return "registration.no_history.restore_backup_failed.message".localized
        }
    }

    fileprivate func showRestoreError(_ error: Error) {
        
        let alert = UIAlertController(title: "registration.no_history.restore_backup_failed.title".localized,
                                      message: errorMessage(for: error),
                                      preferredStyle: .alert)
        
        let tryAgainAction = UIAlertAction(title: "registration.no_history.restore_backup_failed.try_again".localized,
                                          style: .default) { [unowned self] _ in
                                            self.showFilePicker()
        }
        alert.addAction(tryAgainAction)
        
        let cancelAction = UIAlertAction(title: "general.cancel".localized,
                                           style: .cancel) { [unowned self] _ in
            self.formStepDelegate.didCompleteFormStep(self)
        }
        alert.addAction(cancelAction)
        
        self.present(alert, animated: true)
    }
    
    fileprivate func restore(with url: URL) {
        guard let sessionManager = SessionManager.shared else {
            return
        }
        
        self.showLoadingView = true
        sessionManager.restoreFromBackup(at: url) { result in
            switch result {
            case .failure(let error):
                BackupEvent.importFailed.track()
                self.showRestoreError(error)
                self.showLoadingView = false
            case .success:
                BackupEvent.importSucceeded.track()
                self.indicateLoadingSuccessRemovingCheckmark(false) {
                    self.formStepDelegate.didCompleteFormStep(self)
                }
            }
        }
    }
}

extension NoHistoryViewController: UIDocumentPickerDelegate {
    public func documentPicker(_ controller: UIDocumentPickerViewController, didPickDocumentAt url: URL) {
        self.restore(with: url)
    }
}
